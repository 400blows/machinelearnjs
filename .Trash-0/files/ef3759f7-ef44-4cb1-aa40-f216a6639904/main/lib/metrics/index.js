"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const ConfusionMatrix = __importStar(require("ml-confusion-matrix"));
/**
 * Reference: http://www.dataschool.io/simple-guide-to-confusion-matrix-terminology/
 * Official API Doc: https://mljs.github.io/confusion-matrix/
 * @param yTrue
 * @param yPred
 */
exports.confusion_matrix = ConfusionMatrix;
// const trueLabels =      [0, 1, 0, 1, 1, 0];
const trueLabels = [[0, 1, 1], [1, 0, 0]];
const predictedLabels = [1, 1, 1, 1, 0, 0];
const cm2 = new exports.confusion_matrix([[13, 2], [10, 5]], ['cheese', 'dog']);
console.log('acc', cm2.getAccuracy());
console.log(cm2.getFalseCount());
console.log(cm2.getIndex('dog'));
const classification_1 = require("./classification");
const accResult = classification_1.accuracyScore({
    y_true: [0, 1, 2, 3],
    y_pred: [0, 2, 1, 3]
});
console.log('accuracy result ', accResult);
const accResultNorm = classification_1.accuracyScore({
    y_true: [0, 1, 2, 3],
    y_pred: [0, 2, 1, 3],
    normalize: false
});
console.log('accuracy result iwht norm false ', accResultNorm);
const loss_zero_one_result = classification_1.zeroOneLoss({
    y_true: [1, 2, 3, 4],
    y_pred: [2, 2, 3, 5]
});
console.log('loss zero one ', loss_zero_one_result);
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi8uLi8uLi8uLi9zcmMvbGliL21ldHJpY3MvaW5kZXgudHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6Ijs7Ozs7Ozs7O0FBQUEscUVBQXVEO0FBRXZEOzs7OztHQUtHO0FBQ1UsUUFBQSxnQkFBZ0IsR0FBRyxlQUFlLENBQUM7QUFFaEQsOENBQThDO0FBQzlDLE1BQU0sVUFBVSxHQUFHLENBQUMsQ0FBQyxDQUFDLEVBQUUsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsRUFBRSxDQUFDLENBQUMsQ0FBQyxDQUFDO0FBQzFDLE1BQU0sZUFBZSxHQUFHLENBQUMsQ0FBQyxFQUFFLENBQUMsRUFBRSxDQUFDLEVBQUUsQ0FBQyxFQUFFLENBQUMsRUFBRSxDQUFDLENBQUMsQ0FBQztBQUUzQyxNQUFNLEdBQUcsR0FBRyxJQUFJLHdCQUFnQixDQUFDLENBQUMsQ0FBQyxFQUFFLEVBQUUsQ0FBQyxDQUFDLEVBQUUsQ0FBQyxFQUFFLEVBQUUsQ0FBQyxDQUFDLENBQUMsRUFBRSxDQUFDLFFBQVEsRUFBRSxLQUFLLENBQUMsQ0FBQyxDQUFDO0FBRXhFLE9BQU8sQ0FBQyxHQUFHLENBQUMsS0FBSyxFQUFFLEdBQUcsQ0FBQyxXQUFXLEVBQUUsQ0FBQyxDQUFDO0FBQ3RDLE9BQU8sQ0FBQyxHQUFHLENBQUMsR0FBRyxDQUFDLGFBQWEsRUFBRSxDQUFDLENBQUM7QUFDakMsT0FBTyxDQUFDLEdBQUcsQ0FBQyxHQUFHLENBQUMsUUFBUSxDQUFDLEtBQUssQ0FBQyxDQUFDLENBQUM7QUFFakMscURBQThEO0FBRTlELE1BQU0sU0FBUyxHQUFHLDhCQUFhLENBQUM7SUFDOUIsTUFBTSxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsRUFBRSxDQUFDLEVBQUUsQ0FBQyxDQUFDO0lBQ3BCLE1BQU0sRUFBRSxDQUFDLENBQUMsRUFBRSxDQUFDLEVBQUUsQ0FBQyxFQUFFLENBQUMsQ0FBQztDQUNyQixDQUFDLENBQUM7QUFFSCxPQUFPLENBQUMsR0FBRyxDQUFDLGtCQUFrQixFQUFFLFNBQVMsQ0FBQyxDQUFDO0FBRTNDLE1BQU0sYUFBYSxHQUFHLDhCQUFhLENBQUM7SUFDbEMsTUFBTSxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsRUFBRSxDQUFDLEVBQUUsQ0FBQyxDQUFDO0lBQ3BCLE1BQU0sRUFBRSxDQUFDLENBQUMsRUFBRSxDQUFDLEVBQUUsQ0FBQyxFQUFFLENBQUMsQ0FBQztJQUNwQixTQUFTLEVBQUUsS0FBSztDQUNqQixDQUFDLENBQUM7QUFFSCxPQUFPLENBQUMsR0FBRyxDQUFDLGtDQUFrQyxFQUFFLGFBQWEsQ0FBQyxDQUFDO0FBRS9ELE1BQU0sb0JBQW9CLEdBQUcsNEJBQVcsQ0FBQztJQUN2QyxNQUFNLEVBQUUsQ0FBQyxDQUFDLEVBQUUsQ0FBQyxFQUFFLENBQUMsRUFBRSxDQUFDLENBQUM7SUFDcEIsTUFBTSxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsRUFBRSxDQUFDLEVBQUUsQ0FBQyxDQUFDO0NBQ3JCLENBQUMsQ0FBQztBQUVILE9BQU8sQ0FBQyxHQUFHLENBQUMsZ0JBQWdCLEVBQUUsb0JBQW9CLENBQUMsQ0FBQyJ9