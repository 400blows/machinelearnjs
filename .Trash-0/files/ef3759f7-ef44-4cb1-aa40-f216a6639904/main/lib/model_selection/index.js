"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _split_1 = require("./_split");
const X = [[0, 1], [2, 3], [4, 5], [6, 7], [8, 9]];
const y = [0, 1, 2, 3, 4];
console.log('original X', X);
console.log('original y', y);
console.log(_split_1.train_test_split(X, y, {
    test_size: 0.33,
    train_size: 0.67,
    random_state: 42
}));
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiaW5kZXguanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi8uLi8uLi8uLi9zcmMvbGliL21vZGVsX3NlbGVjdGlvbi9pbmRleC50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOztBQUFBLHFDQUE0QztBQUU1QyxNQUFNLENBQUMsR0FBRyxDQUFDLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxFQUFFLENBQUMsQ0FBQyxDQUFDLENBQUM7QUFDbkQsTUFBTSxDQUFDLEdBQUcsQ0FBQyxDQUFDLEVBQUUsQ0FBQyxFQUFFLENBQUMsRUFBRSxDQUFDLEVBQUUsQ0FBQyxDQUFDLENBQUM7QUFFMUIsT0FBTyxDQUFDLEdBQUcsQ0FBQyxZQUFZLEVBQUUsQ0FBQyxDQUFDLENBQUM7QUFDN0IsT0FBTyxDQUFDLEdBQUcsQ0FBQyxZQUFZLEVBQUUsQ0FBQyxDQUFDLENBQUM7QUFDN0IsT0FBTyxDQUFDLEdBQUcsQ0FDVCx5QkFBZ0IsQ0FBQyxDQUFDLEVBQUUsQ0FBQyxFQUFFO0lBQ3JCLFNBQVMsRUFBRSxJQUFJO0lBQ2YsVUFBVSxFQUFFLElBQUk7SUFDaEIsWUFBWSxFQUFFLEVBQUU7Q0FDakIsQ0FBQyxDQUNILENBQUMifQ==