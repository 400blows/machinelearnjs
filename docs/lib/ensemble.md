[kalimdor](../README.md) > [ensemble/RandomForestClassifier](../classes/_lib_ensemble_forest_.randomforestclassifier.md)

# Class: RandomForestClassifier

## Hierarchy

**RandomForestClassifier**

## Index

### Constructors

* [constructor](_lib_ensemble_forest_.randomforestclassifier.md#constructor)

### Methods

* [fit](_lib_ensemble_forest_.randomforestclassifier.md#fit)
* [predict](_lib_ensemble_forest_.randomforestclassifier.md#predict)

---

## Constructors

<a id="constructor"></a>

###  constructor

⊕ **new RandomForestClassifier**(props?: *`object`*): [RandomForestClassifier](_lib_ensemble_forest_.randomforestclassifier.md)

*Defined in [lib/ensemble/forest.ts:6](https://github.com/JasonShin/kalimdorjs/blob/12bcdb8/src/lib/ensemble/forest.ts#L6)*

RandomForestClassifier hey bro

**Parameters:**

| Param | Type | Default value |
| ------ | ------ | ------ |
| `Default value` props | `object` |  { nEstimator: 10 } |

**Returns:** [RandomForestClassifier](_lib_ensemble_forest_.randomforestclassifier.md)

___

## Methods

<a id="fit"></a>

###  fit

▸ **fit**(__namedParameters: *`object`*): `void`

*Defined in [lib/ensemble/forest.ts:21](https://github.com/JasonShin/kalimdorjs/blob/12bcdb8/src/lib/ensemble/forest.ts#L21)*

Build a forest of trees from the training set (X, y).

**Parameters:**

| Param | Type |
| ------ | ------ |
| __namedParameters | `object` |

**Returns:** `void`
void

___
<a id="predict"></a>

###  predict

▸ **predict**(X: *`any`*): `any`[]

*Defined in [lib/ensemble/forest.ts:44](https://github.com/JasonShin/kalimdorjs/blob/12bcdb8/src/lib/ensemble/forest.ts#L44)*

Predict class for X.

The predicted class of an input sample is a vote by the trees in the forest, weighted by their probability estimates. That is, the predicted class is the one with highest mean probability estimate across the trees.

**Parameters:**

| Param | Type | Description |
| ------ | ------ | ------ |
| X | `any` |  array-like or sparse matrix of shape = \[n_samples\] |

**Returns:** `any`[]

___

