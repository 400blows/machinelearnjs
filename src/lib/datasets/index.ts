import { Iris } from './Iris';

export { Iris };
