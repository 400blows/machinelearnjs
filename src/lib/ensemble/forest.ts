import * as _ from 'lodash';
import { DecisionTreeClassifier } from '../tree/tree';

export class RandomForestClassifier {
  private trees = [];
  private nEstimator;

  /**
   *
   * @param {{nEstimator: number}} props
   */
  constructor(
    props: {
      /** test estimator type! */
      nEstimator: number;
    } = {
      /** test estimator default! */
      nEstimator: 10
    }
  ) {
    this.nEstimator = props.nEstimator;
  }

  /**
   * Build a forest of trees from the training set (X, y).
   * @param {Array} X - array-like or sparse matrix of shape = [n_samples, n_features]
   * @param {Array} y - array-like, shape = [n_samples] or [n_samples, n_outputs]
   * @returns void
   */
  public fit({ X, y }): void {
    this.trees = _.reduce(
      _.range(0, this.nEstimator),
      sum => {
        const tree = new DecisionTreeClassifier({
          featureLabels: null,
          randomise: true
        });
        tree.fit({ X, y });
        return _.concat(sum, [tree]);
      },
      []
    );
  }

  /**
   * Predict class for X.
   *
   * The predicted class of an input sample is a vote by the trees in the forest, weighted by their probability estimates.
   * That is, the predicted class is the one with highest mean probability estimate across the trees.
   * @param {Array} X - array-like or sparse matrix of shape = [n_samples]
   * @returns {string[]}
   */
  public predict(X): any[] {
    const predictions = _.map(this.trees, tree => {
      return tree.predict({ X });
    });
    return this.baggingPrediction(predictions);
  }

  /**
   * @hidden
   * Bagging prediction help method
   * According to the predictions returns by the trees, it will select the
   * class with the maximum number (votes)
   * @param {Array<any>} predictions - List of initial predictions that may look like [ [1, 2], [1, 1] ... ]
   * @returns {string[]}
   */
  private baggingPrediction(predictions: any[]): string[] {
    const counts = _.countBy(predictions, x => x);
    const countsArray = _.reduce(
      _.keys(counts),
      (sum, k) => {
        const returning = {};
        returning[k] = counts[k];
        return _.concat(sum, returning);
      },
      []
    );
    const max = _.maxBy(countsArray, x => _.head(_.values(x)));
    const key = _.head(_.keys(max));
    // Returning bagging pred result
    return _.split(key, ',');
  }
}
