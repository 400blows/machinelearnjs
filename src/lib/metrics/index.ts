/* tslint:disable */
import * as ConfusionMatrix from 'ml-confusion-matrix';

/**
 * Reference: http://www.dataschool.io/simple-guide-to-confusion-matrix-terminology/
 * Official API Doc: https://mljs.github.io/confusion-matrix/
 * @param yTrue
 * @param yPred
 */
export const confusion_matrix = ConfusionMatrix;

// const trueLabels =      [0, 1, 0, 1, 1, 0];
const trueLabels = [[0, 1, 1], [1, 0, 0]];
const predictedLabels = [1, 1, 1, 1, 0, 0];

const cm2 = new confusion_matrix([[13, 2], [10, 5]], ['cheese', 'dog']);

console.log('acc', cm2.getAccuracy());
console.log(cm2.getFalseCount());
console.log(cm2.getIndex('dog'));

import { accuracyScore, zeroOneLoss } from './classification';

const accResult = accuracyScore({
  y_true: [0, 1, 2, 3],
  y_pred: [0, 2, 1, 3]
});

console.log('accuracy result ', accResult);

const accResultNorm = accuracyScore({
  y_true: [0, 1, 2, 3],
  y_pred: [0, 2, 1, 3],
  normalize: false
});

console.log('accuracy result iwht norm false ', accResultNorm);

const loss_zero_one_result = zeroOneLoss({
  y_true: [1, 2, 3, 4],
  y_pred: [2, 2, 3, 5]
});

console.log('loss zero one ', loss_zero_one_result);
