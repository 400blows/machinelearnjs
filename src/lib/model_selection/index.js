"use strict";
exports.__esModule = true;
var _split_1 = require("./_split");
var X = [[0, 1], [2, 3], [4, 5], [6, 7], [8, 9]];
var y = [0, 1, 2, 3, 4];
_split_1.train_test_split(X, y);
