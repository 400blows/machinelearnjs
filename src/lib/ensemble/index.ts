import { installTF } from '../utils/deps';
installTF();
import { RandomForestClassifier } from './forest';

export { RandomForestClassifier };
