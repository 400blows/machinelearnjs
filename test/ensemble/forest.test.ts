import { Iris } from '../../src/lib/datasets/Iris';
import { RandomForestClassifier } from '../../src/lib/ensemble/forest';
import { accuracyScore } from '../../src/lib/metrics';
import { train_test_split } from '../../src/lib/model_selection';

describe('ensemble:forest', () => {
  const X1 = [[0, 0], [1, 1], [2, 1], [1, 5], [3, 2]];
  const y1 = [0, 1, 2, 3, 7];
  describe('RandomForest', () => {
    it('should return a result of X1 and y1 with random state 1', () => {
      const randomForest = new RandomForestClassifier({ random_state: 1 });
      randomForest.fit({ X: X1, y: y1 });
      const result = randomForest.predict([[0, 3], [2, 1]]);
      const expected = [1, 2];
      expect(result).toEqual(expected);
    });

    it('should return a result of X1 and y1 with random state 2', () => {
      const randomForest = new RandomForestClassifier({ random_state: 2 });
      randomForest.fit({ X: X1, y: y1 });
      const result = randomForest.predict([[0, 3], [2, 1]]);
      const expected = [0, 2];
      expect(result).toEqual(expected);
    });

    it('should train test split and predict correct test #1', () => {
      const randomForest = new RandomForestClassifier();
      const dataset = new Iris();
      dataset.load();
      const { xTest, xTrain, yTest, yTrain } = train_test_split({ X: dataset.data, y: dataset.targets });
      randomForest.fit({ X: xTrain, y: yTrain });
      const yPred = randomForest.predict(xTest);
      const shouldBeGreater = 0.7;
      const accuracy = accuracyScore({ y_true: yTest, y_pred: yPred });
      expect(accuracy).toBeGreaterThanOrEqual(shouldBeGreater);
    });
  });
});
