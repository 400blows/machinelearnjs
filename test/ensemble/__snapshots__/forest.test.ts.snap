// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for fit function are given 1`] = `[Error: Input cannot be an empty array]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for fit function are given 2`] = `[Error: values passed to tensor(values) must be an array of numbers or booleans, or a TypedArray]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for fit function are given 3`] = `[TypeError: The matrix is not 2D shaped: [1,2] of [2]]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for fit function are given 4`] = `[Error: Input cannot be an empty array]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for fit function are given 5`] = `[TypeError: Number of labels=1 does not math number of samples=2]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for predict function are given 1`] = `[TypeError: The matrix is not 2D shaped: [] of [0]]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for predict function are given 2`] = `[Error: values passed to tensor(values) must be an array of numbers or booleans, or a TypedArray]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for predict function are given 3`] = `[TypeError: The matrix is not 2D shaped: [1,2] of [2]]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for predict function are given 4`] = `[TypeError: Cannot read property 'trees' of undefined]`;

exports[`ensemble:forest RandomForest should throw an exception if incorrect inputs for predict function are given 5`] = `[TypeError: Cannot read property 'trees' of undefined]`;
